8351842: Test issues on Windows with combination of --enable-linkable-runtime and --with-external-symbols-in-bundles=public

RealCLanger opened the pull request:
This PR addresses an issue that can be observed when building on Windows with configure options `--enable-linkable-runtime` and `--with-external-symbols-in-bundles=public`.

The problem is that with this special build configuration, we build two sets of .pdb files for the binaries. The first set is the standard debug symbols files named <binary-name>.pdb. The second set consists of stripped debug symbols file called <binary-name>.stripped.pdb which have less information but enough to present line numbers in hs-err files.

During build we use the *.stripped.pdb files for compiling the jmods and also the bundle files. However, in the images folder, both sets of .pdb files exist. The tests for runtime linking will now, in the absence of jmod files, pick up the .pdb files (without *stripped*) from images, but in the runtime the hashes of the *stripped* files are stored.

With this change, the standard .pdb files in the `--with-external-symbols-in-bundles=public` configuration are now the stripped files and we create a set of full pdb files named *.full.pdb. Jmods and Bundles still contain the stripped pdb files and we also fix the issue that the debug symbols bundle also contained stripped pdb files so far. With this fix, it will contain the full pdb files and extracting these over a JDK runtime will replace stripped pdbs with full pdbs.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8351842](https://bugs.openjdk.org/browse/JDK-8351842): Test issues on Windows with combination of --enable-linkable-runtime and --with-external-symbols-in-bundles=public (**Bug** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24012/head:pull/24012` \
`$ git checkout pull/24012`

Update a local copy of the PR: \
`$ git checkout pull/24012` \
`$ git pull https://git.openjdk.org/jdk.git pull/24012/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24012`

View PR using the GUI difftool: \
`$ git pr show -t 24012`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24012.diff">https://git.openjdk.org/jdk/pull/24012.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24012#issuecomment-2718383674)
</details>

erikj79 commented:
What happens with the local debugging experience in the exploded image if you configure with ` --with-external-symbols-in-bundles=public`? Then the full debug symbols are only present in the *.full.pdb. Will the debugger tools on Windows know automatically how to resolve that, or will this configuration be inconvenient for local dev/debug usecases? As I remember it, the current scheme was chosen to put the "full" pdb files as *.pdb in the exploded image whenever possible.

jerboaa commented:
Nice find! This sounds like a bug in the build machinery.

jerboaa commented:
> What happens with the local debugging experience in the exploded image if you configure with ` --with-external-symbols-in-bundles=public`?

Do we care about the local debugging experience in that case? I'd argue if good debug experience is what is wanted, you'd omit `--with-external-symbols-in-bundles=public` configure option?

erikj79 commented:
> > What happens with the local debugging experience in the exploded image if you configure with ` --with-external-symbols-in-bundles=public`?
> 
> Do we care about the local debugging experience in that case? I'd argue if good debug experience is what is wanted, you'd omit `--with-external-symbols-in-bundles=public` configure option?

I think this can easily be turned around as, do we care about being able to run certain tests using the image in the build directory, rather than using an unpacked bundle, which is what we are actually shipping? The underlying issue here is that the image we produce in the build output directory isn't exactly what we then bundle up and ship. The image is trying to serve multiple purposes and compromises were made a long time ago which we are still carrying today. I'm not sure what the right answer is. We have contemplated enabling `--with-external-symbols-in-bundles=public` for Oracle builds, but with this change, I think it would be a lot less likely to happen, because JDK developers would have to actively opt out of it if we made it default for us.

RealCLanger commented:
First of all, after giving this a spin in our testing I found that I needed to address two new test problems. If we go with this change, we need to align JmodExcludedFiles with the new naming (*.full.pdb instead of *.stripped.pdb), otherwise this test would fail in ` --with-external-symbols-in-bundles=public` mode. Also, with the standard pdb file not being the stripped one, the gtest `dll_address_to_function_and_library_name` fails in its current form because it doesn't print the symbol name any more but only an address. So I weakened the condition on Windows a little to make the test pass. Probably something for the eyes of @tstuefe.

Now, regarding the principal point opened by Erik, I think it is somehow valid. If we were to build with `--with-external-symbols-in-bundles=public`, the debugging experience for a developer is a bit worse. On the other hand, I agree with Severin, developers should probably not use this for their local builds, it's rather a shipment configuration. The alternative would be to exclude some 50+ test cases when using linkable runtime on Windows, which is no good either. Or should there be an alternative image directory used for these kinds of test? Or should jmod get some workaround coding to check whether *stripped.pdb files are present and try to prefer these when building runtimes?

magicus commented:
Please refrain from integrating this until we get consensus on the principal issue here, as pointed out by Erik. This is not just a build issue, it is an issue of differing expectations. It might even be that the solution is a modification in jlink rather than the build system.

magicus commented:
My take is that we should be able to both test and debug what we ship. I can't say straight up how to solve that, but I believe that if there is a will, there is a way; we just have to try harder to find it.

fthevenet commented:
> What happens with the local debugging experience in the exploded image if you configure with ` --with-external-symbols-in-bundles=public`? Then the full debug symbols are only present in the *.full.pdb files. Will the debugger tools on Windows know automatically how to resolve that, or will this configuration be inconvenient for local dev/debug usecases? As I remember it, the current scheme was chosen to put the "full" pdb files as *.pdb in the exploded image whenever possible.

FWIW, from my work on https://github.com/openjdk/jdk/pull/16039 I remember verifying that debuggers on Windows do not rely on the filenames of pdb files to match the symbols and binaries, so AFAIK, the fact that the symbols for, say, `jvm.dll`, will be in a file named `jvm.dll.full.pdb` instead of `Jvm.dll.pdb` does not imply it won't be found by the debugger (i.e. it used to be in a file named `jvm.pdb`)
Now, if I understand the proposed change correctly, the exploded image will contain two distinct `.pdb` files for each binaries, so a question remain about how a debugger might decide to choose one over the other, but do we know for sure that this indeed affect the right symbols to be picked up?

jerboaa commented:
Maybe, I misunderstood how this works... I'm not a Windows person.

> My take is that we should be able to both test and debug what we ship.

Agreed. What I don't understand why this inconsistency is being done to begin with? To appease testing maybe? Is that the idea?

> During build we use the *.stripped.pdb files for compiling the jmods and also the bundle files. However, in the images folder, both sets of .pdb files exist. The tests for runtime linking will now, in the absence of jmod files, pick up the .pdb files (without _stripped_) from images, but in the runtime the hashes of the _stripped_ files are stored.

I'll add here that the implementation of JEP 493, does nothing special. It takes what's in the JMODs *at build time* when `jlink` is being run and uses that as the "source of truth" when it comes to hash sums. It sounds to me that when `jlink` runs during the build it only sees `*.stripped.pdb` files (renamed as `*.pdb`). If something is being done **after** the `jdk` image has been created (by `jlink`) and changes files (like copying full `*.pdb` files over the stripped ones that are actually in the JMODs) then this is the root cause of this bug. Rightfully so, since the `*.pdb` files in the JDK image no longer match the ones in the JMOD files. Is that the case? @RealCLanger Could you please run these commands for me and let me know the output? Sorry, I don't know the windows equivalents of these:

```
$ jmod extract --dir java.base-jmod-extract images\jmods\java.base.jmod
$ sha512sum java.base-jmod-extract\bin\keytool.exe.pdb
5df21e93cd24a03e491debd469001fc73950812de2d2b5d6df0272b55e9d206b56a4332a9f77e19e8755da12dc43f1c353575503b1f53d3d894794e19632964a java.base-jmod-extract\bin\keytool.exe.pdb
$ jimage extract --dir lib-modules-extract images\jdk\lib\modules
$ grep keytool.exe.pdb lib-modules-extract\jdk.jlink\jdk\tools\jlink\internal\runtimelink\fs_java.base_files
6|0|5df21e93cd24a03e491debd469001fc73950812de2d2b5d6df0272b55e9d206b56a4332a9f77e19e8755da12dc43f1c353575503b1f53d3d894794e19632964a|bin\keytool.exe.pdb
$ sha512sum images\jdk\bin\keytool.exe.pdb
<different-to-5df21e93cd24a03e491debd469001fc73950812de2d2b5d6df0272b55e9d206b56a4332a9f77e19e8755da12dc43f1c353575503b1f53d3d894794e19632964a>
```

jerboaa commented:
Another angle of this is that as soon as a user runs `jlink` to produce a custom runtime the debugging experience will be different to the JDK they started out with.

erikj79 commented:
I did a little bit of an experiment. When we link today, we supply this argument to define the main/full pdb file:
```
"-pdb:$$($1_SYMBOLS_DIR)/$$($1_BASENAME).pdb" 
```
And in the case of `--with-external-symbols-in-bundles=public`, we also add this:
```
"-pdbstripped:$$($1_SYMBOLS_DIR)/$$($1_BASENAME).stripped.pdb"
```

Just checking the contents of the resulting dll using `strings`, we can see that a reference to the pdb file is stored:
```
$ strings build/windows-x64/images/jdk/bin/verify.dll | grep pdb
verify.dll.pdb
```

This reference is the same wether I configure with `--with-external-symbols-in-bundles=public` or not, as in, strings does not find any `verify.dll.stripped.pdb` reference. Now with this patch, we instead supply:
```
"-pdb:$$($1_SYMBOLS_DIR)/$$($1_BASENAME).full.pdb" 
"-pdbstripped:$$($1_SYMBOLS_DIR)/$$($1_BASENAME).pdb"
```
And then the stored reference becomes:
```
$ strings build/windows-x64-public-patch/images/jdk/bin/verify.dll | grep pdb
verify.dll.full.pdb
```

This is of course just an indication of what the Microsoft linker does and how the tools will interpret these libraries and executables. I haven't actually tested all the scenarios with debugging tools or generating hserr files. I do think it's enough to warrant a proper investigation before we change the `-pdb` flag to point to a filename other than `$$($1_BASENAME).pdb` though, as that could potentially have consequences for loading any kind of symbols in the final binaries.

fthevenet commented:
I did some tests by building the JDK's master branch (e.g. *without* this patch) with `--with-external-symbols-in-bundles=public` and then renaming the generated `jvm.dll.pdb` to `jvm.dll.full.pdb`. I left `jvm.dll.stripped.pdb` unchanged.

Then using [WinDbg](https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/) to debug jshell, I could confirm that in fact WinDbg could *not* load symbols for `jvm.dll` neither from `jvm.dll.full.pdb` nor from `jvm.dll.stripped.dll`, It would resort to what it calls "exported symbols" for this module, which you can't use for doing much.

Then I renamed  `jvm.dll.full.pdb` back to `jvm.dll.pdb` and restarted the debugger and confirmed I could now set a breakpoint in the source file and have the debugger step in and highlight source lines properly, as expected.

Finally I named  `jvm.dll.stripped.pdb` as  `jvm.dll.full.pdb`  and reloaded everything; this time I could view the functions names in the jvm modules displayed properly in the debugger console when stepping, but could no longer follow its excecution from the source code view. Again this is expected since public symbols contain function names and entry point addresses but not source file and line number info.

So it seems that contrary to what I suggested in my previous comment, it *is* important that the `.pdb` file from where the debugger is expected to load symbols from is the same name as was provided to the linker. 
Now the reason why I initially thought it didn't matter is because `WinDbg` automatically caches pdb files that it successfully loaded under `C:\ProgramData\dbg\sym\` and reload them from there even if the original file no longer exist. So if you load the symbols once, then rename the file and start again, you may be misled into thinking that it successfully loaded the symbols from the renamed file while it in fact did so from its cache.

Hope this helps, and sorry for the potential misleading info earlier.

magicus commented:
Given all of this, is there any possibility of all for --enable-linkable-runtime to handle both full and stripped pdb files, or do we need to accept that this is an inherent limitation for the linkable runtime?

RealCLanger commented:
> Given all of this, is there any possibility at all for --enable-linkable-runtime to handle both full and stripped pdb files, or do we need to accept that this is an inherent limitation for the linkable runtime?

Yes, I think there is 😄  I propose an alternative approach in #24057

jerboaa commented:
> Given all of this, is there any possibility at all for --enable-linkable-runtime to handle both full and stripped pdb files, or do we need to accept that this is an inherent limitation for the linkable runtime?

I see a couple of options:

1. Do the same in JMODs that what is done in the image (*.pdb files are full, *.stripped.pdb are, well stripped). I.e. include them both. The problem should go away. I guess that's not desired for some reason?
2. @RealCLanger work-around for a work-around :) Cannot say, this is a nice solution.
3. Don't do the full *.pdb copy over stripped *.pdb files when building with `--with-external-symbols-in-bundles=public` and cater for the debugging case by providing a symbols bundle which has *.pdb (full) and *.stripped.pdb and copy them over the image.

None of them seem ideal. My preference would be (3), but YMMV.

RealCLanger commented:
I'll close this PR because something in the lines of #24057 looks more promising to me. Let's continue the discussion there.

RealCLanger commented:
/open

MBaesken commented:
> But once we have them both we then have a number of steps in the build (I don't know how many) that could choose to use one or the other, but we seem to lack a way of saying which should be used.

Might make sense to have another configure flag or option for the already existing` '--with-external-symbols-in-bundles=<option>'` to control this (and if this is set, do more or less what Christoph proposed in his PR).

RealCLanger commented:
Hi all, 

I have reopened this PR because I'd like to come back to the proposed coding changes and like to argue again that it seems the best and clearest solution for the issue.

Since the discussion in the PR has grown quite long, let me summarize first: With configure option `--with-external-symbols-in-bundles=public` we build a JDK that contains a set of stripped pdb files (instead of none or full blown pdbs). These stripped pdb files will be packaged into the module binaries and and are then expected by jlink, e.g. known by their hashcodes. We still build full blown pdb files to create the debug symbols bundle from. As it is currently set up, the images folder that is used for testing contains full pdb files named like `library.pdb` and the stripped pdb files named `library.stripped.pdb`. So far, this was no issue but with option `--enable-linkable-runtime` jlink would link runtimes not from jmod files (that contain only the stripped pdb files) but from the image folder directly where we still have the full blown pdb files named `library.pdb`. And those pdb files obviously don't match with the hashsums stored in jlink.

With this proposed change, we would then have stripped pdb files as `library.pdb` and the full pdb files as `library.full.pdb`. I think this is consistent because that's what is actually delivered in the JDK bundle. The reservation against this approach was, though, that it could harm the developer experience because one could not debug directly from the test image any more and would need to first replace the stripped pdb files with the full pdb files.

The question is, how often this problem would really hit a developer. And if so, is it really so hard to restore the full pdb files? I would argue, that having the stripped pdb files in the image directory is more consistent with what we actually ship and hence we better test the actual product. Furthermore, to mitigate this, one could add another make target or some script. However, I think it is definitely better than hacking something into jlink as proposed in the alternative PR #24057. Maybe another solution could be that `--with-external-symbols-in-bundles=public` would become effective in debug builds but only in product builds?

@erikj79 @magicus would you reconsider this?

erikj79 commented:
> With this proposed change, we would then have stripped pdb files as `library.pdb` and the full pdb files as `library.full.pdb`. I think this is consistent because that's what is actually delivered in the JDK bundle. The reservation against this approach was, though, that it could harm the developer experience because one could not debug directly from the test image any more and would need to first replace the stripped pdb files with the full pdb files.
> 
> The question is, how often this problem would really hit a developer. And if so, is it really so hard to restore the full pdb files? I would argue, that having the stripped pdb files in the image directory is more consistent with what we actually ship and hence we better test the actual product. Furthermore, to mitigate this, one could add another make target or some script. However, I think it is definitely better than hacking something into jlink as proposed in the alternative PR #24057. Maybe another solution could be that `--with-external-symbols-in-bundles=public` would become effective in debug builds but only in product builds?

First just for clarity, the current file names are `library.dll.pdb` and `library.dll.stripped.pdb`, just to avoid any confusion in the discussion.

I don't think there is any value in having `library.dll.full.pdb` present in `JDK_IMAGE_DIR`. Instead, I would like to propose that when `--with-external-symbols-in-bundles=public` is set,  we continue building them as `library.dll.pdb` and `library.dll.stripped.pdb` into the exploded image. That will preserve the ability to debug the exploded image. When creating jmods we continue to rename and include the `library.dll.stripped.pdb` files instead of the full pdb files. When creating the image in `JDK_IMAGE_DIR`, we only include the stripped variants from the jmods, named `library.dll.pdb`. The full files we copy manually into the `SYMBOLS_IMAGE_DIR` dir instead, also named `library.dll.pdb`.

The `SYMBOLS_IMAGE_DIR` was introduced quite some time ago, I think by me, and my intention was to long term try to move the symbol files there completely. I never got around to it, it's quite a disruptive change, but I wanted a better 1 to 1 correlation between the images and the bundles we ship. I'm not sure we are ready for that kind of change yet, but I think we can leverage the `SYMBOLS_IMAGE_DIR` for this particular configuration option.

In `RunTests.gmk` we are already looking for PDB files in `SYMBOLS_IMAGE_DIR` and adding any directories containing such files to, and exporting, `_NT_SYMBOL_PATH`. This helps debuggers and symbol lookup on Windows find the the PDB files when running tests through the makefiles. In local runs of `make test` this isn't currently being used, because we never put any PDBs in the `SYMBOLS_IMAGE_DIR`, but in our internal distributed testing (at Oracle), through `make test-prebuilt`, we provide the contents of the symbols bundle as `SYMBOLS_IMAGE_DIR` to get it picked up, without having to overlay the symbols bundle on top of the JDK bundle in our test setup, so we know this works.

With my proposal, full symbols should still be automatically available when running tests on the full JDK image through the makefiles. The only caveat here is if a file `library.dll.pdb` in the same directory as `library.dll` is picked up before anything in `_NT_SYMBOL_PATH`, this would need to be tested. If running tests using Jtreg directly, or debugging other scenarios, the developer would need to opt out of `--with-external-symbols-in-bundles=public` and I think in that case that would be reasonable. I'm mostly after making sure that testing through the makefiles continues to work out of the box.

erikj79 commented on make/CreateJmods.gmk:
```diff
@@ -69,7 +69,7 @@ ifneq ($(LIBS_DIR), )
         DEST := $(LIBS_DIR_FILTERED), \
         FILES := $(FILES_LIBS), \
-        NAME_MACRO := rename_stripped, \
+        NAME_MACRO := rename_stripped \
```
These trailing commas were deliberate and are part of the [Code Conventions for the Build System](https://openjdk.org/groups/build/doc/code-conventions.html) (17, though it's not very clearly stated).

erikj79 commented on make/Images.gmk:
```diff
@@ -282,29 +282,63 @@ else
+  # No space before argument to avoid having to put $(strip ) everywhere in implementation above.
+  $(call SetupCopyStrippedDebuginfo,JDK)
+  $(call SetupCopyStrippedDebuginfo,JRE)
+endif
```
I might be missing something, but why is this needed? Shouldn't the public pdbs already be part of the jmods in this case and so would get linked into the image by jlink?

RealCLanger commented on make/Images.gmk:
```diff
@@ -282,29 +282,63 @@ else
+      $(eval DBGFILES := $(call FindDebuginfoFiles, $(SUPPORT_OUTPUTDIR)/modules_libs/$m)) \
+      $(eval DBGFILES := $(if $(filter true+public,$(call isTargetOs,windows)+$(SHIP_DEBUG_SYMBOLS)), \
+        $(filter-out %.stripped.pdb,$(DBGFILES)),$(DBGFILES)) \
+      ) \
```
Done.

RealCLanger commented:
> From what I can tell, this is also removing the full debug symbols from the actual JDK/JRE images in all cases and all platforms, so they actually match the bundles. I think that's ok, but it needs to be clearly stated somewhere, at least in the bug, for future reference of when that happened.

Yes, you are right. I added a sentence to the JBS bug.

I also implemented your feedback, please look again.
